### PR TITLE
DagsHub Logger: Fix unsupported metric formats for MLflow, Add example notebook

### PR DESCRIPTION
**Pull request by nirbarazida**
MLflow doesn't support some of the formats YOLO-NAS uses, so we either reformat them or raise an expiration that this format is not supported and will not be logged. This way, if YOLO-NAS uses a metric we didn't encounter during our tests and is not supported by MLflow, the training process will not stop, and the user will get an informative error message about it.

### CODE REVIEW

File: src/super_gradients/common/sg_loggers/dagshub_sg_logger.py [hunk 1]
```diff
@@ -165,26 +165,51 @@ def _dvc_add(self, local_path="", remote_path=""):
     def _dvc_commit(self, commit=""):
         self.dvc_folder.commit(commit, versioning="dvc", force=True)
 
+    @multi_process_safe
+    def _get_nested_dict_values(self, d, parent_key="", sep="/"):
+        items = []
+        for k, v in d.items():
+            new_key = parent_key + sep + k if parent_key else k
+            if isinstance(v, dict):
```
Comment (ofrimasad) [inline src/super_gradients/common/sg_loggers/dagshub_sg_logger.py:173]: please use `isinstance(v, Maping)` instead.
some of the dictionaries might be DictConfig

**Review by ofrimasad — COMMENTED** [review]
looks great.
one minor comment

File: src/super_gradients/common/sg_loggers/dagshub_sg_logger.py (hunk 1 above)
Comment (deanp70) [reply src/super_gradients/common/sg_loggers/dagshub_sg_logger.py]: Fixed this comment @ofrimasad sorry for the delayed response.

**Comment by deanp70** [comment]
Hey @ofrimasad, I fixed the issue you mentioned, can this be merged in now?

Thanks,
Dean

File: src/super_gradients/common/sg_loggers/dagshub_sg_logger.py [hunk 2]
```diff
@@ -165,26 +165,51 @@ def _dvc_add(self, local_path="", remote_path=""):
...
-                try:
-                    mlflow.log_params({k: v})
-                except Exception:
-                    logger.warning(f"Skip to log {k}: {v}")
+        flatten_dict = self._get_nested_dict_values(d=config)
+        for k, v in flatten_dict:
+            try:
+                mlflow.log_params({k: v})
+            except Exception as e:
+                logger.debug(e)
```
Comment (BloodAxe) [inline src/super_gradients/common/sg_loggers/dagshub_sg_logger.py:187]: Is there a specific reason why you want to log this with DEBUG severity and not WARNING or ERROR?

File: src/super_gradients/common/sg_loggers/dagshub_sg_logger.py [hunk 3]
```diff
@@ -165,26 +165,51 @@ def _dvc_add(self, local_path="", remote_path=""):
...
+            mlflow.log_metrics(metrics=tag_scalar_dict, step=global_step)
+        except Exception:
+            flatten_dicts = self._get_nested_dict_values(tag_scalar_dict)
+            for k, v in flatten_dicts:
+                try:
+                    if isinstance(v, torch.Tensor):
+                        v = v.item()
+                    else:
+                        v = float(v)
+                    self.add_scalar(tag=k.replace("@", "at"), scalar_value=v, global_step=global_step)
```
Comment (BloodAxe) [inline src/super_gradients/common/sg_loggers/dagshub_sg_logger.py:210]: Is "@" the only character than mlflow does not support in tag name?

File: src/super_gradients/common/sg_loggers/dagshub_sg_logger.py [hunk 4]
```diff
@@ -165,26 +165,51 @@ def _dvc_add(self, local_path="", remote_path=""):
...
+                except Exception as e:
+                    logger.debug(f"error: {e}")
```
Comment (BloodAxe) [inline src/super_gradients/common/sg_loggers/dagshub_sg_logger.py:212]: Let's improve logging messages to include `tag_scalar_dict`.
`"Failed to log {tag_scalar_dict}. Got an exception {e}"` or something like that to give user more information what he tried to log and what didn't work.

File: src/super_gradients/common/sg_loggers/dagshub_sg_logger.py (hunk 2 above)
Comment (BloodAxe) [inline src/super_gradients/common/sg_loggers/dagshub_sg_logger.py:187]: Same here https://github.com/Deci-AI/super-gradients/pull/915/files#r1232049356

File: src/super_gradients/common/sg_loggers/dagshub_sg_logger.py (hunk 3 above)
Comment (timho102003) [reply src/super_gradients/common/sg_loggers/dagshub_sg_logger.py:229]: There are a few more characters that should be checked. I've added the function for checking those unsupported characters and specify in the warning message

File: src/super_gradients/common/sg_loggers/dagshub_sg_logger.py (hunk 2 above)
Comment (timho102003) [reply src/super_gradients/common/sg_loggers/dagshub_sg_logger.py]: Hi @BloodAxe, I've improved the logging messages. Thanks!

**Comment by deanp70** [comment]
@BloodAxe we reviewed and fixed your comments on the unsupported symbols. Please let me know if there's anything else, or if we can go ahead and merge this in. Thanks in advance! 🙏

File: src/super_gradients/common/sg_loggers/dagshub_sg_logger.py [hunk 5]
```diff
@@ -166,28 +167,69 @@ def _dvc_add(self, local_path="", remote_path=""):
...
+                logger.warning(err_msg)
 
     @multi_process_safe
     def add_scalar(self, tag: str, scalar_value: float, global_step: [int, TimeUnit] = 0):
         super(DagsHubSGLogger, self).add_scalar(tag=tag, scalar_value=scalar_value, global_step=global_step)
-        if isinstance(global_step, TimeUnit):
-            global_step = global_step.get_value()
-        mlflow.log_metric(key=tag, value=scalar_value, step=global_step)
+        try:
+            mlflow.log_metric(key=tag, value=scalar_value, step=global_step)
```
Comment (BloodAxe) [inline src/super_gradients/common/sg_loggers/dagshub_sg_logger.py:208]: I feel you may want to have this check here to get integer value of step 
```
if isinstance(global_step, TimeUnit):
   global_step = global_step.get_value()
```

File: src/super_gradients/common/sg_loggers/dagshub_sg_logger.py [hunk 6]
```diff
@@ -166,28 +167,69 @@ def _dvc_add(self, local_path="", remote_path=""):
...
-                except Exception:
-                    logger.warning(f"Skip to log {k}: {v}")
+        flatten_dict = self._get_nested_dict_values(d=config)
+        for k, v in flatten_dict:
+            try:
+                mlflow.log_params({k: v})
+            except Exception as e:
+                is_contain, spec_char = self._contains_special_characters(k)
+                if is_contain:
+                    err_msg = f"Fail to log {k}, please remove the unsupported characters: {spec_char}"
```
Comment (BloodAxe) [inline src/super_gradients/common/sg_loggers/dagshub_sg_logger.py:199]: Hmm, I doubt that is the right thing to do (asking user to fix some parameter, especially if it is coming from SG codebase). 
Maybe it's better to just sanitize the key and remove/replace unsupported characters with "_" or whatever and still log it?
Another option is to support all characters in MLFlow ;) 
WDYT?

File: src/super_gradients/common/sg_loggers/dagshub_sg_logger.py (hunk 5 above)
Comment (deanp70) [reply src/super_gradients/common/sg_loggers/dagshub_sg_logger.py]: Done! Good point

**Comment by deanp70** [comment]
Fixed both comments. Is it possible to merge now?

**Comment by nirbarazida** [comment]
Hi @ofrimasad @BloodAxe 

We've fixed all comments and ran a few tests to check the "happy path" works and other edge cases. 
Is it possible to merge now?

**Comment by nirbarazida** [comment]
Hey @BloodAxe @ofrimasad 
Did you have a chance to review the changes? We want to promote the integration on our channels but waiting for the fixes to be merged.

**Review by BloodAxe — APPROVED** [review]
Sorry, getting this PR took way too long.
LGTM
